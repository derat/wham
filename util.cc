// Copyright 2007 Daniel Erat <dan@erat.org>
// All rights reserved.

#include "util.h"

#include <cstdarg>
#include <cstdlib>
#include <ctime>
#include <iomanip>
#include <sstream>

namespace wham {

Logger::Logger(const string& filename, int line_num)
    : newline_seen_(false),
      orig_format_flags_(cerr.flags()) {
  time_t now = time(NULL);
  struct tm tm;
  localtime_r(&now, &tm);
  char time_str[12];
  strftime(time_str, sizeof(time_str), "%m%d %H%m%S", &tm);
  cerr << time_str << " " << filename << ":" << line_num << "] ";
}

Logger::~Logger() {
  if (!newline_seen_) cerr << "\n";
  cerr.flags(orig_format_flags_);
}

Logger& Logger::operator<<(long v) { cerr << v; return *this; }
Logger& Logger::operator<<(unsigned long v) { cerr << v; return *this; }
Logger& Logger::operator<<(bool v) { cerr << v; return *this; }
Logger& Logger::operator<<(short v) { cerr << v; return *this; }
Logger& Logger::operator<<(unsigned short v) { cerr << v; return *this; }
Logger& Logger::operator<<(int v) { cerr << v; return *this; }
Logger& Logger::operator<<(unsigned int v) { cerr << v; return *this; }
Logger& Logger::operator<<(double v) { cerr << v; return *this; }
Logger& Logger::operator<<(float v) { cerr << v; return *this; }
Logger& Logger::operator<<(long double v) { cerr << v; return *this; }
Logger& Logger::operator<<(const void* v) { cerr << v; return *this; }

Logger& Logger::operator<<(ios_base& (*f)(ios_base&)) {
  cerr << f;
  return *this;
}

Logger& Logger::operator<<(const char* v) {
  return (*this << string(v));
}

Logger& Logger::operator<<(const string& v) {
  newline_seen_ = (!v.empty() && v[v.size()-1] == '\n');
  cerr << v;
  return *this;
}


double GetCurrentTime() {
  struct timeval tv;
  CHECK(gettimeofday(&tv, NULL) == 0);
  return tv.tv_sec + (tv.tv_usec / 1000000.0);
}


void FillTimeval(double time, struct timeval *tv) {
  CHECK(tv);
  tv->tv_sec = static_cast<__time_t>(time);
  tv->tv_usec =
      static_cast<__suseconds_t>(1000000 * (time - static_cast<int>(time)));
}


void SplitString(const string& str, vector<string>* parts) {
  CHECK(parts);
  parts->clear();
  static pcrecpp::RE re("\\s*(\\S+)");
  pcrecpp::StringPiece input(str);
  string part;
  while (re.Consume(&input, &part)) parts->push_back(part);
}


vector<string> SplitString(const string& str) {
  vector<string> parts;
  SplitString(str, &parts);
  return parts;
}


void SplitStringUsing(const string& str,
                      const string& delim,
                      vector<string>* parts) {
  CHECK(parts);
  CHECK(!delim.empty());
  parts->clear();
  size_t start = 0;
  while (start < str.size()) {
    size_t delim_pos = str.find(delim, start);
    if (delim_pos == string::npos) delim_pos = str.size();
    if (delim_pos > start) {
      parts->push_back(str.substr(start, delim_pos - start));
    }
    start = delim_pos + delim.size();
  }
}


string StringPrintf(const char* format, ...) {
  char buffer[1024];  // FIXME: remove magic number?
  va_list argp;
  va_start(argp, format);
  vsnprintf(buffer, sizeof(buffer), format, argp);
  va_end(argp);
  return string(buffer);
}


void JoinString(const vector<string>& parts,
                const string& delim,
                string* output) {
  CHECK(output);
  ostringstream stream;
  bool first = true;
  for (vector<string>::const_iterator part = parts.begin();
       part != parts.end(); ++part) {
    if (!first) stream << delim;
    stream << *part;
    first = true;
  }
  *output = stream.str();
}


string JoinString(const vector<string>& parts, const string& delim) {
  string out;
  JoinString(parts, delim, &out);
  return out;
}

}  // namespace wham
