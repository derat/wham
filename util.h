// Copyright 2007 Daniel Erat <dan@erat.org>
// All rights reserved.

#ifndef __UTIL_H__
#define __UTIL_H__

#include <iostream>
#include <map>
#include <pcrecpp.h>
#include <string>
#include <vector>

using namespace std;

#define LOG wham::Logger(__FILE__, __LINE__)

// FIXME: do something extra here
#define ERROR wham::Logger(__FILE__, __LINE__)
#define DEBUG wham::Logger(__FILE__, __LINE__)

#define CHECK(x) \
  if (!(x)) { \
    wham::Logger(__FILE__, __LINE__) \
        << "Assertion \"" #x "\" failed; exiting"; \
    exit(1); \
  }

#define DISALLOW_EVIL_CONSTRUCTORS(class_name) \
  class_name(const class_name&); \
  void operator=(const class_name&)

#define DEFINE_int(name, default_value) \
  int SETTING_##name = (default_value)

#define DEFINE_string(name, default_value) \
  string SETTING_##name = (default_value)


typedef unsigned int uint;


class UtilTestSuite;

namespace wham {

class Logger {
 public:
  Logger(const string& filename, int line_num);
  ~Logger();

  Logger& operator<<(const string& msg);
  Logger& operator<<(int num);
  Logger& operator<<(void* ptr);
  Logger& operator<<(ios_base& (*f)(ios_base&));

 private:
  // Has the input so far ended with a newline?
  bool newline_seen_;

  // Format flags at the time we were instatiated
  ios_base::fmtflags orig_format_flags_;

  DISALLOW_EVIL_CONSTRUCTORS(Logger);
};  // class Logger


// A reference-counted pointer class.
template<class T>
class ref_ptr {
 public:
  ref_ptr(T* ptr=NULL)
      : ptr_(ptr),
        refs_(ptr ? new int(0) : NULL) {
    add_ref();
  }
  ref_ptr(const ref_ptr<T>& o)
      : ptr_(o.ptr_),
        refs_(o.refs_) {
    add_ref();
  }
  ~ref_ptr() {
    del_ref();
    ptr_ = NULL;
    refs_ = NULL;
  }
  ref_ptr<T>& operator=(const ref_ptr<T>& o) {
    del_ref();
    ptr_ = o.ptr_;
    refs_ = o.refs_;
    add_ref();
    return *this;
  }

  T* get() const {
    return ptr_;
  }
  T &operator*() const {
    return *ptr_;
  }
  T *operator->() const {
    return ptr_;
  }
  void reset(T* ptr=NULL) {
    del_ref();
    ptr_ = ptr;
    refs_ = ptr ? new int(0) : NULL;
    add_ref();
  }
  void swap(ref_ptr<T>& other) {
    T* tmp_ptr = ptr_;
    ptr_ = other.ptr_;
    other.ptr_ = tmp_ptr;

    int* tmp_refs = refs_;
    refs_ = other.refs_;
    other.refs_ = tmp_refs;
  }
  bool operator==(const ref_ptr<T>& other) const {
    return ptr_ == other.ptr_;
  }

 private:
  friend class ::UtilTestSuite;

  inline void add_ref() {
    if (refs_ != NULL) (*refs_)++;
  }
  inline void del_ref() {
    if (refs_ != NULL) {
      (*refs_)--;
      if (*refs_ == 0) {
        delete ptr_;
        delete refs_;
      }
    }
  }

  T* ptr_;
  int* refs_;
};


template<class K, class V>
V FindWithDefault(const map<K, V>& the_map, const K& key, V def) {
  // FIXME: It's dumb that I'm doing two lookups here.  I'm too dense to
  // figure out the syntax for declaring a const_iterator of the templated
  // map class -- the obvious "map<K, V>::const_iterator it" yields
  // "expected `;' before 'it'".
  if (the_map.find(key) == the_map.end()) return def;
  return the_map.find(key)->second;
}


// Split a string on whitespace, saving the individual pieces to the
// passed-in vector.
void SplitString(const string& str, vector<string>* parts);


// Split a string on whitespace, returning the individual pieces as a new
// vector.
vector<string> SplitString(const string& str);


}  // namespace wham

#endif  // __UTIL_H__
