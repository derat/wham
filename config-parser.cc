// Copyright 2007 Daniel Erat <dan@erat.org>
// All rights reserved.

#include "config-parser.h"

#include <iostream>
#include <sstream>
#include <stack>

#include "util.h"

using namespace std;

namespace wham {

string ParsedConfig::Node::Dump(int level) {
  ostringstream out;
  for (int i = 0; i < level; ++i) out << "  ";
  for (vector<string>::const_iterator token = tokens.begin();
       token != tokens.end(); ++token) {
    if (token != tokens.begin()) out << " ";
    // TODO: should be escaping backslashes and doublequotes here
    out << "\"" << *token << "\"";
  }
  out << "\n";
  for (vector<ref_ptr<Node> >::const_iterator child = children.begin();
       child != children.end(); ++child) {
    out << (*child)->Dump(level + 1);
  }
  return out.str();
}


bool ConfigParser::ParseFromFile(const string& filename,
                                 ParsedConfig* config,
                                 vector<string>* errors) {
  CHECK(config);
  FileTokenizer tokenizer(filename);
  return Parse(&tokenizer, config);
}


ConfigParser::Tokenizer::Tokenizer()
    : done_(false),
      line_num_(1),
      ungetted_char_(0),
      have_ungetted_char_(false) {
}


bool ConfigParser::Tokenizer::GetNextToken(
    string* token, TokenType* token_type, bool* error) {
  CHECK(token);
  CHECK(token_type);
  CHECK(error);

  token->clear();
  *error = false;
  if (done_) return false;

  bool in_single_quote = false;
  bool in_double_quote = false;
  bool in_escape = false;
  bool in_comment = false;
  bool in_token = false;
  bool bare_token = true;
  bool bare_last_char = true;

  int quote_start_line = 0;
  int last_ch = 0;
  int ch = 0;

  while (true) {
    last_ch = ch;
    ch = GetChar();

    // First, handle unterminated quoted strings.
    if ((ch == '\n' || ch == EOF) && (in_single_quote || in_double_quote)) {
      ERROR << "Unclosed quoted string started on line " << quote_start_line;
      *error = true;
      return false;
    }

    // If we're in a comment, ignore any characters devoid of meaning
    // within comments.
    if (in_comment && ch != '\n' && ch != EOF) {
      continue;
    }

    // If we saw a whitespace character or EOF and we're not in a quoted
    // string or escape, then we should emit the in-progress token (if
    // there is one).
    if ((isspace(ch) || ch == EOF) &&
        !in_single_quote && !in_double_quote && !in_escape) {
      if (in_token) {
        if (bare_token) {
          *token_type = GetTokenType(*token);
        } else {
          *token_type = TOKEN_LITERAL;
        }
        // These generate their own tokens, so we'll save them for later.
        if ((ch == '\n' && !in_escape) || ch == EOF) UngetChar(ch);
        bare_token = true;
        return true;
      }

      // Ignore non-significant whitespace that didn't signal the end of a
      // token.
      if (ch != '\n' && ch != EOF) continue;
    }

    // If we got a terminating character but didn't return a token in the
    // previous block, then we're seeing the terminator for the second
    // time, so we should return it as a token.
    if (ch == '\n' || ch == EOF) {
      // When we see a newline, increment the line number and terminate
      // comments.
      if (ch == '\n') {
        line_num_++;
        if (in_comment) in_comment = false;
        // Don't return escaped newlines as terminators.
        if (in_escape) {
          in_escape = false;
          continue;
        }
      }
      if (ch == EOF) done_ = true;
      *token = "\n";
      *token_type = TOKEN_NEWLINE;
      return true;
    }

    // Invariant: At this point we won't see any newlines or EOFs, and if
    // there's a whitespace character, it's in a quoted string or part of
    // an escape.
    CHECK(ch != '\n');
    CHECK(ch != EOF);
    CHECK(!isspace(ch) || in_single_quote || in_double_quote || in_escape);

    // Handle single quotes, double quotes, escapes, and comments.
    if (ch == '\'' && !in_escape && !in_double_quote) {
      in_single_quote = !in_single_quote;
      if (!in_token) in_token = true;
      if (in_single_quote) {
        quote_start_line = line_num_;
        bare_token = false;
      }
      continue;
    }
    if (ch == '"' && !in_escape && !in_single_quote) {
      in_double_quote = !in_double_quote;;
      if (!in_token) in_token = true;
      if (in_double_quote) {
        quote_start_line = line_num_;
        bare_token = false;
      }
      continue;
    }
    if (ch == '\\' && !in_escape) {
      in_escape = true;
      continue;
    }
    if (ch == '/' && !in_escape && !in_single_quote && !in_double_quote &&
        bare_last_char && last_ch == '/') {
      in_comment = true;
      *token = token->substr(0, token->size() - 1);
      if (token->empty() && bare_token) in_token = false;
      continue;
    }

    // At this point, we have a character that we're going to add to the
    // in-progress token.
    bare_last_char = true;
    if (in_escape) {
      // Handle escape sequences.
      in_escape = false;
      switch (ch) {
        case 'f': ch = '\f'; break;
        case 'n': ch = '\n'; break;
        case 'r': ch = '\r'; break;
        case 't': ch = '\t'; break;
        case 'v': ch = '\v'; break;
        // Everything else is left untouched.
      }
      bare_token = false;
      bare_last_char = false;
    }
    *token += static_cast<char>(ch);
    in_token = true;
  }
  CHECK(false);
}


bool ConfigParser::Parse(Tokenizer* tokenizer, ParsedConfig* config) {
  CHECK(tokenizer);
  CHECK(config);

  stack<ParsedConfig::Node*> node_stack;
  node_stack.push(&config->root);

  ParsedConfig::Node* current_node = NULL;

  bool in_concat = false;

  string token;
  TokenType token_type = NUM_TOKEN_TYPES;
  bool error = false;
  while (tokenizer->GetNextToken(&token, &token_type, &error)) {
    if (token_type == TOKEN_NEWLINE) {
      // If we're in a concatenation request, we'll ignore the terminator.
      if (in_concat) continue;
      current_node = NULL;
    } else if (token_type == TOKEN_LEFT_BRACE) {
      // FIXME: report an error if we're concatenating
      CHECK(!in_concat);
      // FIXME: create empty current_node if NULL
      CHECK(current_node);
      node_stack.push(current_node);
      current_node = NULL;
    } else if (token_type == TOKEN_RIGHT_BRACE) {
      // FIXME: report an error if we're concatenating
      CHECK(!in_concat);
      node_stack.pop();
      // FIXME: report an error if node_stack is empty
      CHECK(!node_stack.empty());
      current_node = NULL;
    } else {
      // We treat "." tokens as concatenation requests, unless we're
      // already in a request or are at the begin of a node's tokens.
      if (token_type == TOKEN_PERIOD && !in_concat && current_node) {
        in_concat = true;
        continue;
      }
      if (!current_node) {
        current_node = new ParsedConfig::Node;
        node_stack.top()->children.push_back(
            ref_ptr<ParsedConfig::Node>(current_node));
      }
      if (in_concat) {
        // Invariant: if we're concatenating, we already have a token.
        CHECK(current_node && !current_node->tokens.empty());
        current_node->tokens.back().append(token);
        in_concat = false;
      } else {
        current_node->tokens.push_back(token);
      }
    }
  }
  // FIXME: report error if we reach the end while still waiting for a
  // concatenated string
  CHECK(!in_concat);

  return true;
}

}  // namespace wham
