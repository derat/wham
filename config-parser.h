// Copyright 2007 Daniel Erat <dan@erat.org>
// All rights reserved.

#ifndef __CONFIG_PARSER_H__
#define __CONFIG_PARSER_H__

#include <cstdio>
#include <string>
#include <vector>

#include "util.h"

using namespace std;

class ConfigParserTestSuite;

namespace wham {

// An individual node within a config, possibly containing tokens and
// references to sub-nodes.
struct ConfigNode {
  ConfigNode() {}
  string Dump() { return Dump(0); }
  vector<string> tokens;
  vector<ref_ptr<ConfigNode> > children;
  int file_num;  // not yet implemented
  int line_num;
 private:
  string Dump(int level);
  DISALLOW_EVIL_CONSTRUCTORS(ConfigNode);
};


// An error that occurred while loading a config file.
struct ConfigError {
  ConfigError(const string& message, int line_num)
      : file_num(0),
        line_num(line_num),
        message(message) {}

  string ToString() const;

  int file_num;  // not yet implemented
  int line_num;
  string message;
};


// Parses configurations.
class ConfigParser {
 public:
  // Parse a config from 'filename' into 'config'.
  // If 'errors' is non-NULL, errors will be logged there.
  // Returns true on success and false otherwise.
  static bool ParseFromFile(const string& filename,
                            ConfigNode* config,
                            vector<ConfigError>* errors);

 private:
  friend class ::ConfigParserTestSuite;

  enum TokenType {
    TOKEN_LITERAL,
    TOKEN_LEFT_BRACE,
    TOKEN_RIGHT_BRACE,
    TOKEN_PERIOD,
    TOKEN_NEWLINE,
    NUM_TOKEN_TYPES
  };

  // Abstract base class for tokenizing input.
  class Tokenizer {
   public:
    Tokenizer();
    virtual ~Tokenizer() {}

    // Get the next token from the stream.
    // Returns true if a token was returned and false otherwise.
    // The token is written to 'token', its type (relevant in the case of a
    // bare token) is written to 'token_type', the 1-indexed number of the
    // line where it began is written to 'line_num', and 'error' is set if
    // an error was encountered.  'error' only needs to be checked when
    // false is returned.
    bool GetNextToken(
        string* token,
        TokenType* token_type,
        int* line_num,
        bool* error,
        vector<ConfigError>* errors);

   protected:
    // Can the input source be read from?  Returns false if there was an
    // error in its initialization.
    virtual bool Valid() = 0;

    virtual int GetCharImpl() = 0;

   private:
    // Get the next character from the input source.
    int GetChar() {
      if (have_ungetted_char_) {
        have_ungetted_char_ = false;
        return ungetted_char_;
      }
      return GetCharImpl();
    }

    // Push a character that was already read into a temporary buffer such
    // that it will be the next character returned by GetChar().  This
    // buffer can only hold a single character.
    void UngetChar(int ch) {
      CHECK(!have_ungetted_char_);
      ungetted_char_ = ch;
      have_ungetted_char_ = true;
    }

    TokenType GetTokenType(const string& token) {
      if (token == "\n") return TOKEN_NEWLINE;
      if (token == "{")  return TOKEN_LEFT_BRACE;
      if (token == "}")  return TOKEN_RIGHT_BRACE;
      if (token == ".")  return TOKEN_PERIOD;
      return TOKEN_LITERAL;
    }

    bool done_;
    int line_num_;

    int ungetted_char_;
    bool have_ungetted_char_;

    DISALLOW_EVIL_CONSTRUCTORS(Tokenizer);
  };

  // Implementation of Tokenizer that reads from a file.
  class FileTokenizer : public Tokenizer {
   public:
    FileTokenizer(const string& filename)
        : Tokenizer(),
          file_(fopen(filename.c_str(), "r")) {
    }

    ~FileTokenizer() {
      if (file_) {
        fclose(file_);
        file_ = NULL;
      }
    }

   protected:
    bool Valid() {
      return (file_ != NULL);
    }

    int GetCharImpl() {
      CHECK(file_);
      return fgetc(file_);
    }

   private:
    FILE* file_;
  };

  // Implementation of Tokenizer that reads from a string.
  class StringTokenizer : public Tokenizer {
   public:
    StringTokenizer(const string& input)
        : Tokenizer(),
          input_(input),
          pos_(0) {
    }

   protected:
    bool Valid() {
      return true;
    }

    int GetCharImpl() {
      if (pos_ >= input_.size()) return EOF;
      return input_[pos_++];
    }

   private:
    string input_;
    size_t pos_;
  };

  // Parse a config.
  // Returns true on success and false otherwise.
  static bool Parse(
      Tokenizer* tokenizer, ConfigNode* config, vector<ConfigError>* errors);

  DISALLOW_EVIL_CONSTRUCTORS(ConfigParser);
};

}  // namespace wham

#endif
