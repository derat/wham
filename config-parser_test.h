// Copyright 2007 Daniel Erat <dan@erat.org>
// All rights reserved.

#include <cxxtest/TestSuite.h>

#include <cstdarg>
#include <string>
#include <vector>

#include "config-parser.h"
#include "util.h"

using namespace wham;

class ConfigParserTestSuite : public CxxTest::TestSuite {
 public:
  void testTokenizer_GetNextToken_simple() {
    TS_ASSERT(CompareTokens("a b", "a", "b", kNewline, NULL));
    TS_ASSERT(CompareTokens(" a   b  ", "a", "b", kNewline, NULL));
    TS_ASSERT(CompareTokens("a b\n c",
                            "a", "b", kNewline, "c", kNewline, NULL));
    TS_ASSERT(CompareTokens("   ", kNewline, NULL));
    TS_ASSERT(CompareTokens("\n a \n ",
                            kNewline, "a", kNewline, kNewline, NULL));
  }

  void testTokenizer_GetNextToken_special() {
    TS_ASSERT(CompareTokens("a . b", "a", kPeriod, "b", kNewline, NULL));
    TS_ASSERT(CompareTokens("a { b", "a", kLeftBrace, "b", kNewline, NULL));
    TS_ASSERT(CompareTokens("a } b", "a", kRightBrace, "b", kNewline, NULL));
  }

  void testTokenizer_GetNextToken_commenting() {
    TS_ASSERT(CompareTokens("// a b c", kNewline, NULL));
    TS_ASSERT(CompareTokens("a // b c", "a", kNewline, NULL));
    TS_ASSERT(CompareTokens("a//b", "a", kNewline, NULL));
    TS_ASSERT(CompareTokens("//a\nb", kNewline, "b", kNewline, NULL));
    TS_ASSERT(CompareTokens("// a\\nb", kNewline, NULL));
    TS_ASSERT(CompareTokens("a // b\nc", "a", kNewline, "c", kNewline, NULL));
  }

  void testTokenizer_GetNextToken_quoting() {
    TS_ASSERT(CompareTokens("\"a b \"", "a b ", kNewline, NULL));
    TS_ASSERT(CompareTokens("a \"b'c\" d", "a", "b'c", "d", kNewline, NULL));
    TS_ASSERT(CompareTokens("a 'b\"c' d", "a", "b\"c", "d", kNewline, NULL));
    TS_ASSERT(CompareTokens(" '' ", "", kNewline, NULL));
    TS_ASSERT(CompareTokens(" \"\" ", "", kNewline, NULL));
    TS_ASSERT(CompareTokens("\"a\"\"b\"", "ab", kNewline, NULL));
    TS_ASSERT(CompareTokens("\"ab\"'cd'", "abcd", kNewline, NULL));
    TS_ASSERT(CompareTokens("\" // \"", " // ", kNewline, NULL));
    TS_ASSERT(CompareTokens("' // '", " // ", kNewline, NULL));
    TS_ASSERT(CompareTokens("'/'/ blah", "//", "blah", kNewline, NULL));
    TS_ASSERT(CompareTokens("\"\"// blah", "", kNewline, NULL));
    TS_ASSERT(CompareTokens(" \"{\" ", "{", kNewline, NULL));
    TS_ASSERT(CompareTokens(" \'}\' ", "}", kNewline, NULL));
    TS_ASSERT(CompareTokens(" \".\" ", ".", kNewline, NULL));
    TS_ASSERT(CompareTokens(" \"\". ", ".", kNewline, NULL));
  }

  void testTokenizer_GetNextToken_escaping() {
    TS_ASSERT(CompareTokens("\"a \\\" b\"", "a \" b", kNewline, NULL));
    TS_ASSERT(CompareTokens("'a \\' b'", "a ' b", kNewline, NULL));
    TS_ASSERT(CompareTokens(" \\  ", " ", kNewline, NULL));
    TS_ASSERT(CompareTokens("\\\\a\\\\", "\\a\\", kNewline, NULL));
    TS_ASSERT(CompareTokens("\\a", "a", kNewline, NULL));
    TS_ASSERT(CompareTokens(" \\n ", "\n", kNewline, NULL));
    TS_ASSERT(CompareTokens(" \\t ", "\t", kNewline, NULL));
    TS_ASSERT(CompareTokens("a\\nb", "a\nb", kNewline, NULL));
    TS_ASSERT(CompareTokens("a \\\nb", "a", "b", kNewline, NULL));
    TS_ASSERT(CompareTokens("a \\\nn", "a", "n", kNewline, NULL));
    TS_ASSERT(CompareTokens("\\//", "//", kNewline, NULL));
    TS_ASSERT(CompareTokens("a\\///", "a/", kNewline, NULL));
    TS_ASSERT(CompareTokens("\\'abc\\'", "'abc'", kNewline, NULL));
    TS_ASSERT(CompareTokens("\\\"abc\\\"", "\"abc\"", kNewline, NULL));
    TS_ASSERT(CompareTokens("a \\\n b", "a", "b", kNewline, NULL));
    TS_ASSERT(CompareTokens("\\.", ".", kNewline, NULL));
    TS_ASSERT(CompareTokens("\\{", "{", kNewline, NULL));
    TS_ASSERT(CompareTokens("\\}", "}", kNewline, NULL));
  }

  void testTokenizer_GetNextToken_line_num() {
    TS_ASSERT(CompareLineNums("a b\nc d", 1, 1, 1, 2, 2, 2, NULL));
    TS_ASSERT(CompareLineNums("// blah\na", 1, 2, 2, NULL));
    TS_ASSERT(CompareLineNums("\"a b\"", 1, 1, NULL));
  }

  void testConfigParser_parse() {
    ConfigParser::FileTokenizer tokenizer("testdata/config-parser_test.cfg");
    ConfigNode config;
    CHECK(ConfigParser::Parse(&tokenizer, &config));
    string parsed = config.Dump();
    //LOG << parsed;
    string golden;
    CHECK(ReadFileToString("testdata/config-parser_test.golden", &golden));
    TS_ASSERT_EQUALS(parsed, golden);
  }

 private:
  static const char* kLeftBrace;
  static const char* kRightBrace;
  static const char* kPeriod;
  static const char* kNewline;

  bool CompareTokens(const string& input, ...) {
    va_list argp;
    va_start(argp, input);
    vector<string> expected_tokens;
    while (const char* token = va_arg(argp, char*)) {
      expected_tokens.push_back(token);
    }
    va_end(argp);

    ConfigParser::StringTokenizer tokenizer(input);
    string token;
    ConfigParser::TokenType token_type = ConfigParser::NUM_TOKEN_TYPES;
    int line_num = -1;
    bool error = false;
    size_t num_tokens = 0;

    //LOG << "Testing input \"" << input << "\"";
    while (tokenizer.GetNextToken(&token, &token_type, &line_num, &error)) {
      if (error) return false;
      CHECK(num_tokens < expected_tokens.size());
      switch (token_type) {
        case ConfigParser::TOKEN_LEFT_BRACE:  token = kLeftBrace; break;
        case ConfigParser::TOKEN_RIGHT_BRACE: token = kRightBrace; break;
        case ConfigParser::TOKEN_PERIOD:      token = kPeriod; break;
        case ConfigParser::TOKEN_NEWLINE:     token = kNewline; break;
        case ConfigParser::TOKEN_LITERAL:     break;
        default:
          ERROR << "Got unknown token type " << token_type;
          CHECK(false);
      }
      TS_ASSERT_EQUALS(token, expected_tokens[num_tokens]);
      num_tokens++;
    }
    TS_ASSERT_EQUALS(num_tokens, expected_tokens.size());
    return true;
  }

  bool CompareLineNums(const string& input, ...) {
    va_list argp;
    va_start(argp, input);
    vector<int> expected_line_nums;
    while (int line_num = va_arg(argp, int)) {
      expected_line_nums.push_back(line_num);
    }
    va_end(argp);

    ConfigParser::StringTokenizer tokenizer(input);
    string token;
    ConfigParser::TokenType token_type = ConfigParser::NUM_TOKEN_TYPES;
    int line_num = -1;
    bool error = false;
    size_t num_tokens = 0;

    //LOG << "Testing input \"" << input << "\"";
    while (tokenizer.GetNextToken(&token, &token_type, &line_num, &error)) {
      if (error) return false;
      //LOG << "Got " << token << " on line " << line_num;
      CHECK(num_tokens < expected_line_nums.size());
      TS_ASSERT_EQUALS(line_num, expected_line_nums[num_tokens]);
      num_tokens++;
    }
    TS_ASSERT_EQUALS(num_tokens, expected_line_nums.size());
    return true;
  }

  bool ReadFileToString(const string& filename, string* out) {
    FILE* file = fopen(filename.c_str(), "r");
    if (file == NULL) return false;
    char buf[1024];
    while (fgets(buf, sizeof(buf), file)) out->append(buf);
    fclose(file);
    return true;
  }
};

const char* ConfigParserTestSuite::kLeftBrace  = "__LEFT_BRACE__";
const char* ConfigParserTestSuite::kRightBrace = "__RIGHT_BRACE__";
const char* ConfigParserTestSuite::kPeriod     = "__PERIOD__";
const char* ConfigParserTestSuite::kNewline    = "__NEWLINE__";
