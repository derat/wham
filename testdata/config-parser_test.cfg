# This is a full-line comment.
top-level block {
  child.block {
    directive "foo" bar  # trailing comment
  }
  second_block { con . catenation }
  this is a long \
    line
}
"second top-level block" {
  concat .
    enated over a line

  line with "\\" backslashes \\

  escaped \. concatenation "." and more
}
