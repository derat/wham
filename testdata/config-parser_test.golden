
  "top-level" "block"
    "child.block"
      "directive" "foo" "bar"
    "second_block"
      "concatenation"
    "this" "is" "a" "long" "line"
  "second top-level block"
    "concatenated" "over" "a" "line"
    "line" "with" "\" "backslashes" "\"
    "escaped" "." "concatenation" "." "and" "more"
